### DISCUSSION TITLE
Support for bone animations

### DISCUSSION THREAD

**Discussion by skymen** [needle-tools/needle-engine-support#145] [Feature Ideas]
Hey, looking at the examples and the docs, there is no mention of bone animations. Are they supported (even partially)? Or are they planned for a later date?

I know Three has skinned mesh and blending animation support https://threejs.org/examples/?q=animation#webgl_animation_skinning_blending 

but playing around with their demo, the blending kinda sucks compared to the system Unity provides. That was also a massive issue with Luna engine when I used it in the past; the bone animation system was by far the hardest thing to work around when making games using it.

**Comment by marwie** [comment]
Skinned mesh works (just not with custom shader export just now)

For blending animations you can set all up in animatorcontrollers or timeline in Unity

**Comment by marwie** [comment]
Here is a video https://twitter.com/marcel_wiessler/status/1501953954179026946?t=wSvN9qRddXLbTRnt5MaAjg&s=19  and the website uses it in e.g. the "small files" section here: https://needle.tools/?room=filesize

**Comment by skymen** [comment]
https://user-images.githubusercontent.com/11213825/186470916-6e3d238e-bf71-4791-acc6-ea9d430be493.mp4

does it have this blending issue?

Unity has a system for smart blending between animations with "similar" bone movement at different frequencies and it will auto blend the animation speed so both anim "frequencies" match for the whole duration of the transition.

In Three, the wrong blending creates an animation where the character is limping once every few animation loops which looks very bad.

**Comment by skymen** [comment]
This is a big issue IMO because this means we can't do walk->run transitions without running into weird bugs, and it's a scenario a LOT of people are gonna run into, even more so if you want something where the animation speed matches the character's speed (so his feet aren't sliding on the ground) and so you're gonna want to have even finer control over the blending on a longer period of time than a 0.3s transition.

This is where this specific issue becomes a nightmare

**Comment by hybridherbst** [comment]
If you have a specific asset and example where we can look at the issue that would be great.

It does sound more like an issue on three-js side - we can still look at it and decide if we can reasonably fix it / make a PR for it.

Again, a small reproduction project (e.g. works in Unity, looks bad in the web) would be appreciated.

**Comment by skymen** [comment]
I don't have one, I'm just speaking from experience with Threejs and Luna (that also does "exporting to the web from Unity")

For a simple repro, just take a walk and a run anim from mixamo, put it on a character, and blend between both animations using a slow transition. I don't have time to set up Unity/Needle this week, but if you still want me to make a repro project for it next week I'll try to do it
